File: backend/central-api/src/main/java/com/cybelinx/platform/api/security/AuthorizationService.java
package com.cybelinx.platform.api.security;

import com.cybelinx.platform.api.persistence.TenantMembershipRepository;
import java.util.LinkedHashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Port of {@code AuthorizationService}: flattens ACTIVE memberships into role/permission access
 * entries for a user.
 */
@Service
public class AuthorizationService {

    public static final String PLATFORM_ADMIN_ROLE = "CYBELINX_PLATFORM_ADMIN";

    private final TenantMembershipRepository memberships;

    public AuthorizationService(TenantMembershipRepository memberships) {
        this.memberships = memberships;
    }

    /** Port of {@code AuthorizationService.listAccess()}. */
    @Transactional(readOnly = true)
    public List<PlatformAccess> listAccess(UUID userId) {
        Map<UUID, AccessAccumulator> byMembership = new LinkedHashMap<>();

        for (Object[] row : memberships.listMembershipRoles(userId)) {
            UUID membershipId = (UUID) row[0];
            UUID tenantId = (UUID) row[1];
            String roleCode = (String) row[2];
            byMembership.computeIfAbsent(membershipId, id -> new AccessAccumulator(membershipId, tenantId)).roles.add(roleCode);
        }

        for (Object[] row : memberships.listMembershipPermissions(userId)) {
            UUID membershipId = (UUID) row[0];
            String permissionCode = (String) row[1];
            AccessAccumulator acc = byMembership.get(membershipId);
            if (acc != null) {
                acc.permissions.add(permissionCode);
            }
        }

        return byMembership.values().stream()
                .map(acc -> new PlatformAccess(acc.membershipId, acc.tenantId, List.copyOf(acc.roles), List.copyOf(acc.permissions)))
                .toList();
    }

    @Transactional(readOnly = true)
    public boolean isPlatformAdmin(UUID userId) {
        return listAccess(userId).stream().anyMatch(entry -> entry.roles().contains(PLATFORM_ADMIN_ROLE));
    }

    @Transactional(readOnly = true)
    public boolean hasPermission(UUID userId, String permission) {
        return listAccess(userId).stream()
                .anyMatch(entry -> entry.roles().contains(PLATFORM_ADMIN_ROLE) || entry.permissions().contains(permission));
    }

    /** Port of {@code PlatformAccess}. */
    public record PlatformAccess(UUID membershipId, UUID tenantId, List<String> roles, List<String> permissions) {}

    private static final class AccessAccumulator {
        private final UUID membershipId;
        private final UUID tenantId;
        private final Set<String> roles = new LinkedHashSet<>();
        private final Set<String> permissions = new LinkedHashSet<>();

        private AccessAccumulator(UUID membershipId, UUID tenantId) {
            this.membershipId = membershipId;
            this.tenantId = tenantId;
        }
    }
}

File: retired/ts-backend/central-api/src/modules/identity/authorization.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';

export interface PlatformAccess {
  tenantId: string;
  membershipId: string;
  roles: string[];
  permissions: string[];
}

@Injectable()
export class AuthorizationService {
  constructor(private readonly prisma: PrismaService) {}

  async listAccess(userId: string): Promise<PlatformAccess[]> {
    const memberships = await this.prisma.tenantMembership.findMany({
      where: { userId, status: 'ACTIVE' },
      include: {
        memberRoles: {
          include: {
            role: {
              include: {
                rolePermissions: { include: { permission: true } },
              },
            },
          },
        },
      },
    });

    return memberships.map((membership) => {
      const roles = Array.from(
        new Set(membership.memberRoles.flatMap((grant) => (grant.role ? [grant.role.code] : []))),
      );
      const permissions = Array.from(
        new Set(
          membership.memberRoles.flatMap((grant) =>
            grant.role ? grant.role.rolePermissions.map((granted) => granted.permission.code) : [],
          ),
        ),
      );
      return { tenantId: membership.tenantId, membershipId: membership.id, roles, permissions };
    });
  }

  async isPlatformAdmin(userId: string): Promise<boolean> {
    const access = await this.listAccess(userId);
    return access.some((entry) => entry.roles.includes('CYBELINX_PLATFORM_ADMIN'));
  }

  async hasPermission(userId: string, permission: string): Promise<boolean> {
    const access = await this.listAccess(userId);
    return access.some(
      (entry) => entry.roles.includes('CYBELINX_PLATFORM_ADMIN') || entry.permissions.includes(permission),
    );
  }
}